mlir/lib/Dialect/Rock/Transforms/LowerRockReduce.cpp

```diff
@@ -0,0 +1,233 @@
+  auto dataPerThread = (elementCount + (totalThreads - 1)) / totalThreads;
+
+  llvm::errs() << "totalThreads = " << totalThreads << "\n";
+  llvm::errs() << "elementCount = " << elementCount << "\n";
+  llvm::errs() << "dataPerThread = " << dataPerThread << "\n";
+
+  SmallVector<SmallString<8>> topNames;
+  SmallVector<int64_t> topSizes;
+  int64_t topSizeProduct = 1;
+  SmallVector<unsigned int> topDims;
+  SmallVector<StringRef> topNameRefs;
+  for (auto dimAndSize : llvm::enumerate(inpShape.getShape())) {
+    auto dim = dimAndSize.index();
+    auto dimSize = dimAndSize.value();
+    SmallString<8> mname(Twine("m" + Twine(dim)).str());
+    topNames.push_back(mname);
+    topNameRefs.push_back(topNames.back());
+    topSizes.push_back(dimSize);
+    topSizeProduct *= dimSize;
+    topDims.push_back(dim);
+  }
+
+  BottomUpTMBuilder threadsToInpTensor(rewriter, topNameRefs, topSizes, loc);
+  threadsToInpTensor.merge("flatDim", 0, topNameRefs);
+  TransformMapAttr threadsToInpTensorAttr = threadsToInpTensor.get();
+  redInput =
+      rewriter.create<TransformOp>(loc, redInput, threadsToInpTensorAttr);
+
+  threadsToInpTensor =
+      BottomUpTMBuilder::above(threadsToInpTensor, threadsToInpTensorAttr);
+  threadsToInpTensor.pad({"flatDim"},
+                         {0, totalThreads * dataPerThread - topSizeProduct});
+  threadsToInpTensorAttr = threadsToInpTensor.get();
+  redInput =
+      rewriter.create<TransformOp>(loc, redInput, threadsToInpTensorAttr);
+
+  threadsToInpTensor =
+      BottomUpTMBuilder::above(threadsToInpTensor, threadsToInpTensorAttr);
+  threadsToInpTensor.unmerge({"bid", "tid", "iter"}, {0, 1, 2}, "flatDim",
+                             {gridSize, blockSize, dataPerThread});
+  TransformMapAttr tidIterAttr = threadsToInpTensor.get();
+  Value transformed = rewriter.create<TransformOp>(loc, redInput, tidIterAttr);
+
+  return transformed;
+}
+
+static LogicalResult getStoreMethod(ReduceMethod rMethod,
+                                    StoreMethod &stMethod) {
+  if (rMethod == ReduceMethod::Sum) {
+    stMethod = StoreMethod::AtomicAdd;
+    return success();
+  }
+  return failure();
+}
+
+LogicalResult ReduceRewritePattern::matchAndRewrite(
+    ReduceOp op, ReduceOpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+  auto loc = op->getLoc();
+  uint64_t redAxis = adaptor.getAxisAttr().getInt();
+  auto gridSize = adaptor.getGridSizeAttr().getInt();
+  auto blockSize = adaptor.getBlockSizeAttr().getInt();
+
+  // This pass doing this rewrite means the preceding
+  // fusion passes did not perform any fusion and the reduction
+  // op will become the anchor op. Therefore, setting the
+  // blockSize and gridSize attr of the func.
+  auto funcOp = op->getParentOfType<func::FuncOp>();
+  if (funcOp->hasAttr("block_size")) {
+    llvm::errs() << "blockSize = " << funcOp->getAttr("block_size") << "\n";
+    return op.emitError("func op should not have a blockSize set for a "
```

Comment:
Also, strong disagree. It's a kernel, give it a block size and grid size attribute just like the rest of them. This breaks the uniformity of `kernel`

Reply:
That is, we set the attributes on the function at the time we determine them.

Reply:
Hmmm.... currently these are set in AffixTuningParams.
Do you mean to move it there ?

Reply:
Or for this case, we determine them at creation (TosaToRock) -- maybe there then ?

Reply:
The ugly thing is that there's no function there. 

What I'd do is to the copy over in affix-tuning-params.

Reply:
Yeah, I mean this copy from the attributes to the surrounding function should happen, say, in affix-tuning-params, not in this lowering pattern.

Reply:
pushed tosa to rock conversions.